File: sdk/typescript/src/http.ts
import {
  AIIdentityError,
  AuthenticationError,
  ForbiddenError,
  NotFoundError,
  RateLimitError,
  ValidationError,
} from "./errors.js";

const ERROR_MAP: Record<number, new (msg: string, code: string) => AIIdentityError> = {
  401: AuthenticationError,
  403: ForbiddenError,
  404: NotFoundError,
  422: ValidationError,
  429: RateLimitError,
};

/** Build a query string from params, stripping undefined values. */
function toQueryString(params: Record<string, unknown>): string {
  const entries = Object.entries(params).filter(
    ([, v]) => v !== undefined && v !== null,
  );
  if (entries.length === 0) return "";
  const qs = entries
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`)
    .join("&");
  return `?${qs}`;
}

export class HttpClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private readonly timeout: number;

  constructor(apiKey: string, baseUrl: string, options?: { timeout?: number }) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.timeout = options?.timeout ?? 30_000;
    this.headers = {
      "X-API-Key": apiKey,
      "Content-Type": "application/json",
      "User-Agent": "ai-identity-typescript/0.1.0",
    };
  }

  private async request<T>(
    method: string,
    path: string,
    options?: { json?: unknown; params?: Record<string, unknown> },
  ): Promise<T> {
    const qs = options?.params ? toQueryString(options.params) : "";
    const url = `${this.baseUrl}${path}${qs}`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: this.headers,
        body: options?.json ? JSON.stringify(options.json) : undefined,
        signal: controller.signal,
      });

      if (!response.ok) {
        await this.handleError(response);
      }

      if (response.status === 204) return undefined as T;
      return (await response.json()) as T;
    } finally {
      clearTimeout(timer);
    }
  }

  private async handleError(response: Response): Promise<never> {
    const status = response.status;
    let errorCode = "unknown_error";
    let message = `HTTP ${status}`;

    try {
      const body = await response.json();
      if (body?.error) {
        errorCode = body.error.code ?? errorCode;
        message = body.error.message ?? message;
      } else if (body?.detail) {
        message = String(body.detail);
      }
    } catch {
      message = (await response.text().catch(() => "")) || message;
    }

    const ErrorClass = ERROR_MAP[status] ?? AIIdentityError;
    throw new ErrorClass(message, errorCode);
  }

  async get<T>(path: string, params?: Record<string, unknown>): Promise<T> {
    return this.request<T>("GET", path, { params });
  }

  async post<T>(path: string, json?: unknown): Promise<T> {
    return this.request<T>("POST", path, { json });
  }

  async put<T>(path: string, json?: unknown): Promise<T> {
    return this.request<T>("PUT", path, { json });
  }

  async patch<T>(path: string, json?: unknown): Promise<T> {
    return this.request<T>("PATCH", path, { json });
  }

  async delete<T>(path: string): Promise<T> {
    return this.request<T>("DELETE", path);
  }
}

File: sdk/python/src/ai_identity/_http.py
"""Low-level HTTP client for the AI Identity API."""

from __future__ import annotations

from typing import Any

import httpx

from ai_identity.exceptions import (
    AIIdentityError,
    AuthenticationError,
    ForbiddenError,
    NotFoundError,
    RateLimitError,
    ValidationError,
)

_ERROR_MAP: dict[int, type[AIIdentityError]] = {
    401: AuthenticationError,
    403: ForbiddenError,
    404: NotFoundError,
    422: ValidationError,
    429: RateLimitError,
}


class HTTPClient:
    """Async HTTP client that injects auth and maps errors."""

    def __init__(
        self,
        *,
        api_key: str,
        base_url: str,
        timeout: float = 30.0,
    ) -> None:
        self._client = httpx.AsyncClient(
            base_url=base_url,
            headers={
                "X-API-Key": api_key,
                "Content-Type": "application/json",
                "User-Agent": "ai-identity-python/0.1.0",
            },
            timeout=timeout,
        )

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: dict | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Send a request and return the parsed JSON response."""
        # Strip None values from query params
        if params:
            params = {k: v for k, v in params.items() if v is not None}

        response = await self._client.request(
            method,
            path,
            json=json,
            params=params,
        )

        if response.status_code >= 400:
            self._raise_error(response)

        if response.status_code == 204:
            return None

        return response.json()

    def _raise_error(self, response: httpx.Response) -> None:
        """Parse the API error response and raise a typed exception."""
        status = response.status_code
        error_code = "unknown_error"
        message = f"HTTP {status}"

        try:
            body = response.json()
            if "error" in body:
                error_code = body["error"].get("code", error_code)
                message = body["error"].get("message", message)
            elif "detail" in body:
                message = str(body["detail"])
        except Exception:
            message = response.text or message

        exc_class = _ERROR_MAP.get(status, AIIdentityError)
        raise exc_class(message, status_code=status, error_code=error_code)

    async def get(self, path: str, *, params: dict[str, Any] | None = None) -> Any:
        return await self.request("GET", path, params=params)

    async def post(self, path: str, *, json: dict | None = None) -> Any:
        return await self.request("POST", path, json=json)

    async def put(self, path: str, *, json: dict | None = None) -> Any:
        return await self.request("PUT", path, json=json)

    async def patch(self, path: str, *, json: dict | None = None) -> Any:
        return await self.request("PATCH", path, json=json)

    async def delete(self, path: str) -> Any:
        return await self.request("DELETE", path)

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._client.aclose()
